sandman/source/schedule.cpp

```diff
@@ -26,42 +21,116 @@
 // Types
 //
 
-// A schedule event.
-struct ScheduleEvent
-{
-	// Read a schedule event from JSON. 
-	//
-	// p_EventObject:	The JSON object representing the event.
-	//	
-	// Returns:		True if the event was read successfully, false otherwise.
-	//
-	bool ReadFromJSON(rapidjson::Value::ConstObject const& p_EventObject);
-	
-	// Delay in seconds before this entry occurs (since the last).
-	unsigned int	m_DelaySec;
-	
-	// The control action to perform at the scheduled time.
-	ControlAction	m_ControlAction;
-};
-
 // Locals
 //
 
 // Whether the system is initialized.
 static bool s_ScheduleInitialized = false;
 
-// This will contain the schedule once it has been parsed in.
-static std::vector<ScheduleEvent> s_ScheduleEvents;
-
 // The current spot in the schedule.
 static unsigned int s_ScheduleIndex = UINT_MAX;
 
 // The time the delay for the next event began.
 static Time s_ScheduleDelayStartTime;
 
+static Schedule *s_Schedule = NULL;
+
 // Functions
 //
 
+// Schedule members
+
+Schedule::Schedule() {
+}
+
+void Schedule::AddEvent(ScheduleEvent event)
+{
+	m_ScheduleEvents.push_back(event);
+}
+
+bool Schedule::IsEmpty()
+{
+	return (m_ScheduleEvents.size() == 0);
+}
+
+size_t Schedule::GetNumEvents() 
+{
+	return m_ScheduleEvents.size();
+}
+
+std::vector<ScheduleEvent> Schedule::GetEvents()
+{
+	return m_ScheduleEvents;
+}
+
+Schedule *Schedule::CreateSchedule(const char *jsonFileName)
+{
+	Schedule *l_Schedule = NULL;
+
+	auto* l_ScheduleFile = fopen(jsonFileName, "r");
+
+	if (l_ScheduleFile == nullptr)
+	{
+		LoggerAddMessage("Failed to open the schedule file %s.\n", jsonFileName);
+		return NULL;
+	}
+
+	static constexpr unsigned int l_ReadBufferCapacity = 65536;
+	char l_ReadBuffer[l_ReadBufferCapacity];
+	rapidjson::FileReadStream l_ScheduleFileStream(l_ScheduleFile, l_ReadBuffer, 
+		sizeof(l_ReadBuffer));
+
+	rapidjson::Document l_ScheduleDocument;
+	l_ScheduleDocument.ParseStream(l_ScheduleFileStream);
+
+	if (l_ScheduleDocument.HasParseError() == true)
+	{
+		LoggerAddMessage("Failed to parse the schedule file %s.\n", jsonFileName);
+		fclose(l_ScheduleFile);
+		return NULL;
+	}
+
+	// Find the list of events.
+	auto const l_EventsIterator = l_ScheduleDocument.FindMember("events");
+
+	if (l_EventsIterator == l_ScheduleDocument.MemberEnd())
+	{
+		LoggerAddMessage("No schedule events in %s.\n", jsonFileName);
+		fclose(l_ScheduleFile);
+		return NULL;		
+	}
+
+	if (l_EventsIterator->value.IsArray() == false)
+	{
+		fclose(l_ScheduleFile);
+		LoggerAddMessage("No event array in %s.\n", jsonFileName);
+		return NULL;
+	}
+
+	l_Schedule = new Schedule();
```

Comment:
I have generally been avoiding using new. Partially this is just videogame habits, but it does mean we are less prone to all of the associated memory allocation issues. Because it doesn't look like you can load more than one schedule at a time, it probably makes sense to have this be more of a load in place to get around this issue.